herbert_core/utilities/maths/noisify.m

```diff
@@ -17,40 +17,75 @@
+%   regardless of whether e is present, so the first overload output is
+%   probably incomplete.
 
-if nargin==3 && ischar(varargin{1})
-    if is_stringmatchi(varargin{1},'poisson')
-        yout=zeros(size(y));
-        for i=1:numel(y)
-            yout(i)=randpoisson(abs(y(i)));
-        end
-        eout=abs(y);  % the inpuy y is the mean and variance of the Poisson distribution
-    else
-        error('Unrecognised random sampling distribution')
+% deal with optional arguments
+is_poisson = false;
+fac = 0.1;
+USE_LOCAL_MAX = -999; % set default ymax to distinctive negative value
+if nargin>=2 % avoids the case where the only paramemter is y; for >=2 e will not be optional
+    [fac, is_poisson, ymax] = parse_args(y, e, fac, is_poisson, USE_LOCAL_MAX, varargin{:});
+end
+
+% Use Poisson distribution and ignore other arguments
+if is_poisson
+    yout=zeros(size(y));
+    for i=1:numel(y)
+        yout(i)=randpoisson(abs(y(i)));
     end
+    eout=abs(y);  % the input y is the mean and variance of the Poisson distribution
 else
-    if nargin==3
-        if isnumeric(varargin{1}) && isscalar(varargin{1})
-            fac = varargin{1};
-        else
-            error('Noise as fraction of peak signal must be real scalar')
-        end
-    else
-        fac = 0.1;
+    % if ymax was not set by an argument, set from max of |y|
+    if ymax == USE_LOCAL_MAX
+        ymax = max(abs(y(:)));
     end
-    ymax = max(abs(y(:)));          % find maximum magnitude of y for arbitrary dimensions
+
+    % make noise dy and add to y for output; make error bar for noise    
     dy=(fac*ymax)*randn(size(y));   % st. dev. of randn is sigma=1
     yout=y+dy;
     eout=ones(size(y))*(fac*ymax)^2;
 end
 
-if exist('e','var') && isempty(e)
+% adds e (the input variance) to eout if it exists 
+% (it may not,see @sqw/nosify)
+if exist('e','var') && ~isempty(e)
     if isequal(size(e),size(eout))
-        eout=sqrt(eout+e.^2);
+        eout=eout+e;
     else
-        error('Input array of error bars must have same size as input y array')
+        error('HERBERT:noisify', 'Input array of error bars must have same size as input y array')
+    end
+end
+end
+
+function [fac, is_poisson, ymax] = parse_args(y, e, fac, is_poisson, use_local_max, varargin)
+    p = inputParser;
+    addRequired(p, 'y', @isnumeric);   % y compulsory
+    addRequired(p, 'e', @isnumeric);   % e compulsory if any of remaining optional/parameter arguments present
+    numeric_or_poisson = @(x) isnumeric(x) || strcmpi(x,'poisson');
+    addOptional(p, 'dist_or_factor', fac, numeric_or_poisson);  % fac (numeric) or distribution ('poisson') optional, default to fac=0.1
+    addParameter(p,'maximum_value', use_local_max, @isnumeric);  % ymax, as parameter 'maximum_value'. Default to highly visible named negative value.
+    parse(p,y,e,varargin{:});
+
+    % pick up signal max value as either default or input
+    ymax = p.Results.maximum_value;
```

Comment:
Rather than pass a magic number as the maxval in as an argument, you could define that as a local (function level) magic number and move the    
```
if ymax == USE_LOCAL_MAX
        ymax = max(abs(y(:)));
end
```
block into this function.

Then the caller will have the correct value for maximum without needing any additional logic.

Reply:
Pushed to do this. Now changed flag to -inf after consideration of existence of negative y values as per previous conversations.